src/plonk/lookup.rs

```diff
@@ -65,244 +52,155 @@ use crate::{
+        let lhs_rel = h * (l + r.clone()) - Expression::Constant(F::ONE);
         // check gi(ti+r)-mi=0 or check (ti+r)*(gi(ti+r)-mi)=0 for perfect completeness
-        let t_rel = (g * (t + r) - m).expand();
-        Self {
-            k1,
-            k2,
-            t: t_vec,
-            t_commitment,
-            l_rel,
-            t_rel,
-        }
+        let rhs_rel = g * (t + r) - m;
+        (lhs_rel, rhs_rel)
+    }
+
+    /// collect lhs of log-derivative relations from all lookup arguments
+    pub fn log_derivative_lhs(&self, cs: &ConstraintSystem<F>) -> Vec<Expression<F>> {
+        let cha_index = if self.has_vector_lookup { 1 } else { 0 };
+        (0..self.num_lookups())
+            .map(|lookup_index| self.log_derivative_expr(cs, lookup_index, cha_index).0)
+            .collect()
+    }
+
+    /// evaluate each of the lookup expressions to get vector l_i
+    /// where l_i = L_i(x_1,...,x_a)
+    pub fn evaluate_ls(&self, table: &TableData<F>, r: F) -> Vec<Vec<F>> {
+        let data = LookupEvalDomain { table, r };
+        let nrow = 2usize.pow(table.k);
+        self.lookup_polys
+            .iter()
+            .map(|expr| expr.expand())
+            .map(|poly| {
+                (0..nrow)
+                    .into_par_iter()
+                    .map(|row| poly.eval(row, &data))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>()
     }
-}
 
-impl<F: PrimeField> Witness<F> {
+    /// evaluate each of the table expressions to get vector t_i
+    /// where t_i = T(y1,...,y_b)
+    pub fn evaluate_ts(&self, table: &TableData<F>, r: F) -> Vec<Vec<F>> {
+        let data = LookupEvalDomain { table, r };
+        let nrow = 2usize.pow(table.k);
+        self.table_polys
+            .iter()
+            .map(|expr| expr.expand())
+            .map(|poly| {
+                (0..nrow)
+                    .into_par_iter()
+                    .map(|row| poly.eval(row, &data))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>()
+    }
     /// calculate the coefficients {m_i} in the log derivative formula
     /// m_i = sum_j \xi(w_j=t_i) assuming {t_i} have no duplicates
-    pub fn log_derivative_coeffs<C: CurveAffine<ScalarExt = F>>(
-        &self,
-        ss: &Structure<C>,
-    ) -> Vec<F> {
+    pub fn evaluate_m(&self, l: &[F], t: &[F]) -> Vec<F> {
         let mut m: Vec<F> = Vec::new();
         let mut processed_t = Vec::new();
-        for t_i in &ss.t {
+        for t_i in t {
             if processed_t.contains(&t_i) {
                 // If the current t_i has already been processed, push 0 to m
                 m.push(F::ZERO);
             } else {
-                let m_i = self.l.iter().filter(|l_j| *l_j == t_i).count() as u128;
+                let m_i = l.iter().filter(|l_j| *l_j == t_i).count() as u128;
                 m.push(F::from_u128(m_i));
                 processed_t.push(t_i);
             }
```

Comment:
While I was understanding why there are no optimisations here (sorted `Vec` or `HashSet`), I refactored it a little bit

```rust
pub fn evaluate_m(&self, l: &[F], t: &[F]) -> Vec<F> {
    let mut processed_t = Vec::new();
                                                                                 
    t.iter()
        .map(|t_i| {
            if processed_t.contains(&t_i) {
                F::ZERO
            } else {
                processed_t.push(t_i);
                F::from_u128(l.iter().filter(|l_j| *l_j == t_i).count() as u128)
            }
        })
        .collect()
}
```